cmake/modules/AddSwift.cmake

```diff
@@ -466,8 +466,10 @@ function(_add_swift_runtime_link_flags target relpath_to_lib_dir bootstrapping)
       # This requires to explicitly add all the needed compatibility libraries. We
       # can take them from the current build.
       set(vsuffix "-${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}-${SWIFT_HOST_VARIANT_ARCH}")
-      set(conctarget "swiftCompatibilityConcurrency${vsuffix}")
-      target_link_libraries(${target} PUBLIC ${conctarget})
+      foreach(compat_library IN ITEMS "Concurrency" "DynamicReplacements" "50" "51")
+        set(conctarget "swiftCompatibility${compat_library}${vsuffix}")
+        target_link_libraries(${target} PUBLIC ${conctarget})
+      endforeach()
```

Comment:
I think that rather than using `foreach` here, we should be explicitly spelling this out.  The build rules really should be generally declarative:

```
  target_link_libraries(${target} PUBLIC
    swiftCompatibilityConcurrency${vsuffix}
    swiftCompatibilityDynamicReplacements${vsuffix}
    swiftCompatibility50${vsuffix}
    swiftCompatibility51${vsuffix})
```

I think that hoisting this out a bit might be useful.  Perhaps we can get away with something like this instead?

```
add_library(swiftCompatibility INTERFACE)
target_link_libraries(swiftCompatibility INTERFACE
  swiftCompatibilityConcurrency${vsuffix}
  swiftCompatibilityDynamicReplacements${vsuffix}
  swiftCompatibility50${vsuffix}
  swiftCompatibility51${vsuffix})
```

This would mean that we define the set once and re-use it to avoid people forgetting a missing compatibility shim.

Reply:
So there used to be a `_link_built_compatibility_libs` function in `AddSwift` that would take a target and link all of the compatibility libraries against it and set the link directories, etc. That was removed here; https://github.com/apple/swift/pull/40370

The old definition was;
```
function(_link_built_compatibility_libs executable)
   set(platform ${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR})
   target_link_directories(${executable} PRIVATE
     ${SWIFTLIB_DIR}/${platform})
   add_dependencies(${executable}
     "swiftCompatibility50-${platform}"
     "swiftCompatibility51-${platform}"
     "swiftCompatibilityDynamicReplacements-${platform}")
 endfunction()
```

Bringing something like this back (maybe not setting this search path) so that there is only one place to update might be beneficial. Thoughts?

Reply:
I think that the use of functions is a less desirable approach.  The interface library creates a placeholder name, which then propagates the dependencies and is more declarative rather than imperative for handling the dependencies (you just link that one library).  We can also inject any necessary flags (e.g. library search paths) through that for linking against the compatibility shims.

Reply:
I took a stab in 02faa56a1a27b78a095e3fa397bddba9832af8d4 to incorporate this and Evan's feedback in the main discussion -- this will likely need some polishing (e.g. names and style), at this point I would like to check if this approach goes in the right direction.

Reply:
Turns out I had to iterate more since -- see https://github.com/apple/swift/pull/60728#pullrequestreview-1085798384